x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts

```diff
@@ -477,114 +472,32 @@ export class ObservabilityAIAssistantClient {
-            this.dependencies.actionsClient.execute({
-              actionId: connectorId,
-              params: subAction,
-            })
-          )
-        ).pipe(
-          switchMap((executeResult) => {
-            if (executeResult.status === 'error' && executeResult?.serviceMessage) {
-              const tokenLimitRegex =
-                /This model's maximum context length is (\d+) tokens\. However, your messages resulted in (\d+) tokens/g;
-              const tokenLimitRegexResult = tokenLimitRegex.exec(executeResult.serviceMessage);
-
-              if (tokenLimitRegexResult) {
-                const [, tokenLimit, tokenCount] = tokenLimitRegexResult;
-                throw createTokenLimitReachedError(
-                  parseInt(tokenLimit, 10),
-                  parseInt(tokenCount, 10)
-                );
-              }
-            }
-
-            if (executeResult.status === 'error') {
-              throw createInternalServerError(
-                `${executeResult?.message} - ${executeResult?.serviceMessage}`
-              );
+  ): Observable<ChatCompletionChunkEvent | TokenCountEvent | ChatCompletionMessageEvent> => {
+    const tools = functions?.reduce((acc, fn) => {
+      acc[fn.name] = {
+        description: fn.description,
+        schema: fn.parameters,
+      };
+      return acc;
+    }, {} as Record<string, { description: string; schema: any }>);
+
+    const chatComplete$ = defer(() =>
+      this.dependencies.inferenceClient.chatComplete({
+        connectorId,
+        stream: true,
+        messages: convertMessagesForInference(
+          messages.filter((message) => message.message.role !== MessageRole.System)
+        ),
+        functionCalling: simulateFunctionCalling ? 'simulated' : 'native',
+        toolChoice: functionCall
+          ? {
+              function: functionCall,
             }
-
-            const response = executeResult.data as Readable;
-
-            signal.addEventListener('abort', () => response.destroy());
-
-            return tracer.startActiveSpan(
-              '/chat/completions',
-              {
-                kind: SpanKind.CLIENT,
-              },
-              ({ span }) => {
-                return adapter.streamIntoObservable(response).pipe(
-                  withLangtraceChatCompleteSpan({
-                    span,
-                    messages,
-                    functions,
-                    model: connector.name,
-                    serviceProvider:
-                      connector.actionTypeId === ObservabilityAIAssistantConnectorType.OpenAI
-                        ? LangtraceServiceProvider.OpenAI
-                        : LangtraceServiceProvider.Anthropic,
-                  })
-                );
-              }
-            );
-          })
-        );
-      }),
+          : ToolChoiceType.auto,
+        tools,
+      })
+    ).pipe(
+      convertInferenceEventsToStreamingEvents(),
```

Comment:
@dgieselaar 
convertInferenceEventsToStreamingEvents converts any InferenceChatCompletionEvent types to their corresponding StreamingChatResponseEvent types, such as ChatCompletionChunkEvent, TokenCountEvent, ensuring we maintain compatibility with existing event types. This keeps the ObservabilityAIAssistantClient.chat method’s signature intact and avoids exposing inference-client-specific types directly.

Reply:
++, this looks good!